jina/types/document/graph.py

```diff
@@ -0,0 +1,339 @@
+    ):
+        """
+        Add an edge to the graph connecting `doc1` with `doc2`
+
+        :param doc1: the starting node for this edge
+        :param doc2: the ending node for this edge
+        :param features: Optional features dictionary to be added to this new created edge
+        """
+        from scipy.sparse import coo_matrix
+
+        for doc in [doc1, doc2]:
+            self.add_node(doc)
+
+        current_adjacency = self.adjacency
+        doc1_node_offset = self._node_id_to_offset[doc1.id]
+        doc2_node_offset = self._node_id_to_offset[doc2.id]
+        row = (
+            np.append(current_adjacency.row, doc1_node_offset)
+            if current_adjacency is not None
+            else np.array([doc1_node_offset])
+        )
+        col = (
+            np.append(current_adjacency.col, doc2_node_offset)
+            if current_adjacency is not None
+            else np.array([doc2_node_offset])
+        )
+        data = (
+            np.append(current_adjacency.data, 1)
+            if current_adjacency is not None
+            else np.array([1])
+        )
+        self.adjacency = coo_matrix((data, (row, col)))
+        if features is not None:
+            self.edge_features[f"{doc1.id}-{doc2.id}"] = features
+
+    def _remove_edge_id(self, edge_id: int, edge_feature_key: str):
+        from scipy.sparse import coo_matrix
+
+        if self.adjacency is not None:
+            if edge_id > self.num_edges:
+                raise Exception(
+                    f"Trying to remove edge {edge_id} while number of edges is {self.num_edges}"
+                )
+            row = np.delete(self.adjacency.row, edge_id)
+            col = np.delete(self.adjacency.col, edge_id)
+            data = np.delete(self.adjacency.data, edge_id)
+            if row.shape[0] > 0:
+                self.adjacency = coo_matrix((data, (row, col)))
+            else:
+                self.adjacency = coo_matrix((0, 0))
+
+            if edge_feature_key in self.edge_features:
+                del self.edge_features[edge_feature_key]
+
+    def remove_edge(self, doc1: "Document", doc2: "Document"):
+        """
+        Remove a node from the graph along with the edges that may contain it
+
+        :param doc1: the starting node for this edge
+        :param doc2: the ending node for this edge
+        """
+        offset1 = self._node_id_to_offset[doc1.id]
+        offset2 = self._node_id_to_offset[doc2.id]
+        for edge_id, (row, col) in enumerate(
+            zip(self.adjacency.row, self.adjacency.col)
+        ):
+            if row.item() == offset1 and col.item() == offset2:
+                self._remove_edge_id(edge_id, f"{doc1.id}-{doc2.id}")
+
+    @property
+    def edge_features(self):
+        """
+        The dictionary of edge features, indexed by `edge_id` in the `edge list`
+
+        .. # noqa: DAR201
+        """
+        return StructView(self._pb_body.graph_info.edge_features)
```

Comment:
i suggest we rename `graph_info` to `graph` in protobuf:
```suggestion
        return StructView(self._pb_body.graph.edge_features)
```

Reply:
As u prefer